app/views/landing_page/landing_page.erb

```diff
@@ -178,14 +178,25 @@
       ReactDOM.render(React.createElement(componentFn, props), document.getElementById('topbar-placeholder'));
     };
 
-    static_props = <%= topbar_props.to_json.html_safe %>
-    marketplaceContext = <%= marketplace_context.to_json.html_safe %>
-
-    renderTopBar(static_props, marketplaceContext);
-
-    window.fetch("<%= topbar_props_path %>", { credentials: 'same-origin' })
-      .then(function (res) { return res.json(); })
-      .then(function (props) { renderTopBar(Object.assign({}, static_props, props), marketplaceContext); });
+    staticProps = <%= topbar_props.to_json.html_safe %>
+    staticMarketplaceContext = <%= marketplace_context.to_json.html_safe %>
+
+    var p = Promise.race([
+      window.fetch("<%= topbar_props_path %>", { credentials: 'same-origin' }),
+      new Promise(function (resolve, reject) {
+        setTimeout(function() { reject(new Error('request timeout')) }, 3000);
+      })]);
+
+    p.then(function (resJson) {
+      props = Object.assign({}, staticProps, resJson.props);
+      mCtx = Object.assign({}, staticMarketplaceContext, resJson.marketplaceContext)
+      renderTopBar(props, mCtx);
+    });
+
+    p.catch(function(error) {
+      console.log(error);
+      renderTopBar(staticProps, staticMarketplaceContext);
+    });
   </script>
```

Comment:
You are creating multiple global variables, e.g. `p` that pollute all the JS scope. You could wrap the whole script tag in an IIFE:

```html
<script>
  (function () {
    // ...
  }());
</script>
```

This creates a scope around the code.